### A basic guide to working with Dates and Times in Python

There are three relevant packages that can be used to do datetime operations in python.
- `dateutil`
- `datetime`
- `time`

The first two packages are included with the standard python distribution. 
The first package is a third-party package and you can install it using command `pip install dateutil`. For most use cases, last two packages are enough and we'll be focusing more on those. 

Let's get started!

In [1]:
# import the required packages
import time
import datetime
import dateutil

I'm assuming you are familiar with a UNIX timestamp format. The easiest way to calculate time between any to events is using time packages. A simple demonstration follows:-

In [4]:
# Get current timestamp
time.time()

1533306161.8891551

In [6]:
start_time = time.time()
time.sleep(5)
end_time = time.time()
print('Time taken: {:.4f}s'.format(end_time - start_time))

Time taken: 5.0038s


#### `datetime` package:

But just the pure timestamp like `1533306161.8891551` is not very human friendly. The time represented by this timestamp is not obvious to us directly. We are more familiar with our traditional time represenatation formats. This is where the `datetime` package comes in.

The `datetime` package represents the time in the form of a `datetime` object. This object look something like this -
    
    datetime.datetime(2018, 8, 3, 19, 56, 35, 444627)
    
A legend for the values given in the `datetime` object is as -

    datetime(year, month, day, hour, minute, second, microsecond)
    
But this object is still not that user-friendly. 
What if you want to represent dates in the same form as you write on your notebook or include in your letters?

Let's explore some functions of this `datetime` object.

In [9]:
# To get the datetime object for the current time instance, we do the following
datetime.datetime.now()

datetime.datetime(2018, 8, 3, 20, 1, 25, 386361)

In [12]:
# But dates you want to represent are not always the current time instances.
# In this case, you can directly pass the values to the datetime constructor according to the above format.

datetime.datetime(2018, 8, 3, 10, 2, 24, 123455)

datetime.datetime(2018, 8, 3, 10, 2, 24, 123455)

In [14]:
# You don't even have to supply all the values. The constructor will assume the values where missing.
# The following object represents the datetime object for the time instance of 3rd Aug 2018
datetime.datetime(2018, 8, 3)

datetime.datetime(2018, 8, 3, 0, 0)

In [17]:
# What if you want to convert a date or time represented as 03/08/2018 into a datetime object directly?

datetime.datetime.strptime('03/08/2018', '%d/%m/%Y')

datetime.datetime(2018, 8, 3, 0, 0)

We just specified the format of the input string and the constructor will parse the string for us automatically.

Don't worry about the 2nd parameter value, we will get to it. Just know that -

    %d/%m/%Y ---> day/month/year(full year) e.g 03/08/2018, 04/06/2015
    %d-%m-%Y ---> day-month-year(full year) e.g. 03-08-2018, etc
    
You can do a similar thing for the time as well. What if you want to read time from a format like this-

    03-08-2018 23:34:90

In [22]:
l = datetime.datetime.strptime('03-08-2018 23:34:45', '%d-%m-%Y %H:%M:%S')
l

datetime.datetime(2018, 8, 3, 23, 34, 45)

##### `datetime` operations:

But why the hell are we converting our dates into `datetime` objects? What's the benefit?

In [27]:
# The most obvious benefit is that you can print the date and time in any format you want.
# NOTE: We are using the function strptime for reading the date and time.
#       And we are using the function strftime for printing the date and time.

print(l.strftime('%d/%m/%Y'))
print(l.strftime('%d-%m-%Y'))
print(l.strftime('%H:%M:%S'))
print(l.strftime('%d H %m M %S S'))

03/08/2018
03-08-2018
23:34:45
03 H 08 M 45 S


In [35]:
# You can use this to convert the date from one format to another

input_fmt = '2018-04-30 9hr 45min'
print('%-15s : %-20s' % ('Input format', input_fmt))
d = datetime.datetime.strptime(input_fmt, '%Y-%m-%d %Hhr %Mmin')
print('%-15s : %-20s' % ('Datetime format', d.__repr__()))
custom_fmt = d.strftime('%d-%m-%Y %H:%M:%S')
print('%-15s : %-20s' % ('New format', custom_fmt))

Input format    : 2018-04-30 9hr 45min
Datetime format : datetime.datetime(2018, 4, 30, 9, 45)
New format      : 30-04-2018 09:45:00 


Not just for converting the values and formats of date and time. `datetime` objects also support arithmetic opertations.

#### (optional)  you can use the `dateutil` package to make creating `datetime` objects even easier.

In [41]:
d1 = dateutil.parser.parse('08/06/2017')
d2 = dateutil.parser.parse('07/06/2017')
print(d1, d2)

2017-08-06 00:00:00 2017-07-06 00:00:00


In [ ]:
# And lastly, you can create a datetime object directly from the timestamp as well
timestmap = time.time()
print('%-15s : %-20s' % ('Timestamp', time))
